src/main/java/org/opensearch/neuralsearch/search/collector/HybridTopFieldDocSortCollector.java

```diff
@@ -0,0 +1,411 @@
+    // visibility, then anyone will be able to extend the class, which is not what
+    // we want.
+    public HybridTopFieldDocSortCollector(final int numHits, final HitsThresholdChecker hitsThresholdChecker) {
+        this.numHits = numHits;
+        this.hitsThresholdChecker = hitsThresholdChecker;
+    }
+
+    // Add the entry in the Priority queue
+    private void add(int slot, int doc, FieldValueHitQueue<FieldValueHitQueue.Entry> compoundScore, int subQueryNumber, float score) {
+        FieldValueHitQueue.Entry bottomEntry = new FieldValueHitQueue.Entry(slot, docBase + doc);
+        bottomEntry.score = score;
+        bottom = compoundScore.add(bottomEntry);
+        // The queue is full either when totalHits == numHits (in SimpleFieldCollector), in which case
+        // slot = totalHits - 1, or when hitsCollected == numHits (in PagingFieldCollector this is hits
+        // on the current page) and slot = hitsCollected - 1.
+        assert slot < numHits;
+        queueFull[subQueryNumber] = slot == numHits - 1;
+    }
+
+    private void updateBottom(int doc, FieldValueHitQueue<FieldValueHitQueue.Entry> compoundScore) {
+        bottom.doc = docBase + doc;
+        bottom = compoundScore.updateTop();
+    }
+
+    private boolean canEarlyTerminate(Sort searchSort, Sort indexSort) {
+        return canEarlyTerminateOnDocId(searchSort) || canEarlyTerminateOnPrefix(searchSort, indexSort);
+    }
+
+    private boolean canEarlyTerminateOnDocId(Sort searchSort) {
+        final SortField[] fields1 = searchSort.getSort();
+        return SortField.FIELD_DOC.equals(fields1[0]);
+    }
+
+    private boolean canEarlyTerminateOnPrefix(Sort searchSort, Sort indexSort) {
+        if (indexSort != null) {
+            final SortField[] fields1 = searchSort.getSort();
+            final SortField[] fields2 = indexSort.getSort();
+            // early termination is possible if fields1 is a prefix of fields2
+            if (fields1.length > fields2.length) {
+                return false;
+            }
+            return Arrays.asList(fields1).equals(Arrays.asList(fields2).subList(0, fields1.length));
+        } else {
+            return false;
+        }
+    }
+
+    public List<TopFieldDocs> topDocs(final FieldValueHitQueue<FieldValueHitQueue.Entry>[] compoundScores, final Sort sort) {
+        if (compoundScores == null) {
+            return new ArrayList<>();
+        }
+        final List<TopFieldDocs> topFieldDocs = IntStream.range(0, compoundScores.length)
+            .mapToObj(
+                i -> topDocsPerQuery(
+                    0,
+                    Math.min(collectedHits[i], compoundScores[i].size()),
+                    compoundScores[i],
+                    collectedHits[i],
+                    sort.getSort()
+                )
+            )
+            .collect(Collectors.toList());
+        return topFieldDocs;
+    }
+
+    public abstract class HybridTopDocSortLeafCollector implements LeafCollector {
```

Comment:
This class being a public class can be moved out of this file to a separate file with correct access modifiers

Reply:
Actually no, this class is operates on the state of the collector by using instance variable. If we think of moving this class out then the constructor of the class will become super heavy as more than 10-12 variables. Also functionality might break. What I can do is I can make this class protected.

Reply:
The code therefore is become difficult to maintain. That is why keeping in the same class is recommended.